torch/csrc/jit/codegen/cuda/lower_utils.cpp

```diff
@@ -277,6 +337,195 @@ std::pair<kir::ForLoop*, int64_t> getAllocPoint(
 }
 
 } // namespace loop_utils
+
+namespace kir_utils {
+
+namespace {
+
+class ReplaceExprInput : public kir::MutableIrVisitor {
+ public:
+  static kir::Expr* replace(
+      kir::Expr* expr,
+      const std::unordered_map<kir::Val*, kir::Val*>& replacement_map) {
+    ReplaceExprInput replacer(expr, replacement_map);
+    expr->accept(&replacer);
+    TORCH_INTERNAL_ASSERT(expr != nullptr);
+    TORCH_INTERNAL_ASSERT(replacer.replaced_expr_ != nullptr);
+    auto ret_expr = replacer.replaced_expr_;
+
+    // Copy predicates if the original expr is predicated
+    if (ret_expr != expr) {
+      ret_expr->setPredicate(expr->predicate());
+    }
+    return ret_expr;
+  }
+
+  static std::vector<kir::Expr*> replace(
+      const std::vector<kir::Expr*>& scope,
+      const std::unordered_map<kir::Val*, kir::Val*>& replacement_map) {
+    std::vector<kir::Expr*> ret_expr;
+    ret_expr.reserve(scope.size());
+
+    for (auto expr : scope) {
+      ret_expr.push_back(replace(expr, replacement_map));
+    }
+
+    return ret_expr;
+  }
+
+ private:
+  ReplaceExprInput(
+      kir::Expr* expr,
+      const std::unordered_map<kir::Val*, kir::Val*>& replacement_map)
+      : gpu_lower_(GpuLower::current()),
+        ir_builder_(gpu_lower_->kernel()),
+        replacement_map_(replacement_map) {
+    replaced_expr_ = expr;
+  }
+
+  c10::optional<std::unordered_map<kir::Val*, kir::Val*>>
+  getMaybeInputReplacementMap(kir::Expr* expr) {
+    bool need_replacement = false;
+
+    std::unordered_map<kir::Val*, kir::Val*> replaced_val;
+    for (auto in : expr->inputs()) {
+      auto replace_it = replacement_map_.find(in);
+      if (replace_it != replacement_map_.end()) {
+        need_replacement = true;
+        replaced_val[in] = replace_it->second;
+      } else {
+        replaced_val[in] = in;
+      }
+    }
+    if (need_replacement) {
+      return c10::optional<std::unordered_map<kir::Val*, kir::Val*>>(
+          replaced_val);
+    } else {
+      return c10::nullopt;
+    }
+  }
+
+  // IR visitor interface
+  void visit(kir::ForLoop* for_loop) final {
+    auto new_for_loop = ir_builder_.create<kir::ForLoop>(
+        for_loop->iter_domain(),
+        for_loop->index(),
+        for_loop->start(),
+        for_loop->stop(),
+        for_loop->step(),
+        for_loop->vectorize(),
+        for_loop->vectorize_shift());
```

Comment:
This looks error-prone. Let's create a clone function that only copies the loop attributes without the body. Something like:

```
kir::ForLoop* kir::ForLoop::cloneWithoutBody() {
   return ir_builder.create<kir::ForLoop>(...);
}
```

Reply:
I would really like to have that on kernel IR, but in this PR I was trying not to let infrastructure improvement dominate the code changes. But we could discuss a bit here and maybe I can still add some smaller improvements.

Reply:
For this case I guess I could just add a few functions in lower_utils on kernel IR, and we could combine them if we do decide to have a ir_cloner for kernel IR as well.